### PR TITLE
feat: add type hints for param_store.py in `params` module

### PR DESCRIPTION
**Pull request by willtai**
This PR adds type hints to the `params` module. This address [this a section of this issue to add type hints](https://github.com/pyro-ppl/pyro/issues/2550).

I intend to add more type hints wherever I can in future PRs.

### CODE REVIEW

**Review by eb8680 — CHANGES_REQUESTED** [review]
@willtai thanks for the PR! Can you run `make lint` locally, fix any linter warnings, then run `make format`?

File: pyro/params/param_store.py [hunk 1]
```diff
@@ -4,10 +4,13 @@
 import re
 import warnings
 import weakref
+from collections.abc import Iterator, KeysView
```
Comment (eb8680) [inline pyro/params/param_store.py:7]: @willtai I think you need to replace `collections.abc.Iterator` with `typing.Iterable` for tests to pass

Comment (willtai) [reply pyro/params/param_store.py]: thanks! However, I imported `typing.Iterator` rather than `typing.Iterable`

**Review by fritzo — COMMENTED** [review]
This is great!  Could you also remove the `[mypy-pyro.params.*]` section from [setup.cfg](https://github.com/pyro-ppl/pyro/blob/dev/setup.cfg)?  We have been using setup.cfg as an ignore-list while we progressively add type hints to Pyro's codebase.

**Comment by ordabayevy** [comment]
@willtai are you still working on this? Would it be okay if I take over this PR?

**Comment by fritzo** [comment]
Has anyone tried vscode's new copilot "brushes" feature to add type hints?  I've bee having good luck just telling gpt-3.5-turbo to refactor my python code, and we may be able to automate much of this type hinting.

**Comment by ordabayevy** [comment]
Replaced by #3271

**Pull request closed without merging**